[5.0][RFC][Events] Make captcha plugins use plugin events

Fedik opened the pull request:
Pull Request for Issue #28800 .

### Summary of Changes

The patch improve Captcha plugins.
Now it can use a **real events,** and people stop confusing `onDisplay` (etc) callback with event.

**How this work.**
Here added a new event `onCaptchaSetup` and an Interface for `CaptchaProvider` class.
While this event the plugin should register own `CaptchaProvider` class in `CaptchaRegistry`.
The system will retrieve the requested Captcha type from that registry and call required methods, that defined by `CaptchaProviderInterface` 

This is fully backward compatible.

Looking for feedback. 
If all good we can do the same for Editor and its Buttons.

*Techicaly now one plugin can provide multiple captcha, however for now the global configuration done per plugin.
But this can be extended in future, when need.*

### Testing Instructions

Apply patch, remove `administrator/cache/autoload_psr4.php`
Enable `recaptcha_invisible` and make sure it works.

Testing B/C:
Enable older `recaptcha` and make sure it works.

### Actual result BEFORE applying this Pull Request
All works

### Expected result AFTER applying this Pull Request
All works


### Link to documentations
Please select:
- [x] Documentation link for docs.joomla.org: https://docs.joomla.org/Plugin/Events
- [x] Pull Request link for manual.joomla.org: https://github.com/joomla/Manual/pull/96

Fedik committed:
CaptchaProviderInterface, CaptchaRegistry, CaptchaEvent

Fedik committed:
CaptchaNotFoundException

Fedik committed:
CaptchaRegistry service provider

Fedik committed:
Update Captcha class

Fedik committed:
InvisibleCaptchaProvider

Fedik committed:
Comment

Fedik committed:
Folder

dgrammatiko commented:
@Fedik could you also move the captcha plugin to service providers, like https://github.com/joomla/joomla-cms/pull/39639

Fedik commented:
Sure, will update later.

laoneo commented:
> @Fedik could you also move the captcha plugin to service providers, like #39639

I would do this in a separate pr to not block each other.

dgrammatiko commented:
> Sure, will update later.

Probably leave that for another PR. I'm waiting for the Editors, EditorsXtd version

Fedik committed:
phpcs

Fedik committed:
Sync. Merge branch '4.3-dev' into captcha-event2

Fedik committed:
Only values

Fedik committed:
Retrieve name from instance

Fedik committed:
Make sure registry has been init before add

Fedik committed:
Sync:  Merge branch '4.3-dev' into captcha-event2

 Conflicts:
	plugins/captcha/recaptcha_invisible/recaptcha_invisible.xml
	plugins/captcha/recaptcha_invisible/src/Extension/InvisibleReCaptcha.php

Fedik committed:
Update

Fedik committed:
Update

Fedik committed:
Update

Fedik committed:
Update

Fedik committed:
Update

Fedik committed:
Update

dgrammatiko commented on plugins/captcha/recaptcha_invisible/src/Extension/InvisibleReCaptcha.php:
```diff
@@ -28,201 +27,62 @@
 
-            return false;
-        }
+        $result[] = array(
+            Text::_('PLG_CAPTCHA_RECAPTCHA_INVISIBLE') => array(
+                Text::_('PLG_RECAPTCHA_INVISIBLE_PRIVACY_CAPABILITY_IP_ADDRESS'),
+            ),
+        );
```
```suggestion
        $result[] = [
            Text::_('PLG_CAPTCHA_RECAPTCHA_INVISIBLE') => [
                Text::_('PLG_RECAPTCHA_INVISIBLE_PRIVACY_CAPABILITY_IP_ADDRESS'),
            ],
        ];
```

Fedik committed:
Update plugins/captcha/recaptcha_invisible/src/Extension/InvisibleReCaptcha.php

Co-authored-by: Dimitris Grammatikogiannis <dg@dgrammatiko.dev>

laoneo commented:
Why do you don't you just use plugin events and plugins, instead of having a registry, etc.? From my understanding I just want to dispatch an event and the result contains the captcha. Why all these extra classes? I would also remove the Captcha class as it is only an unnecessary static singleton.

Fedik committed:
Update

Fedik commented:
> Why do you don't you just use plugin events and plugins

That what I am doing

>  From my understanding I just want to dispatch an event and the result contains the captcha.

No, that wrong. You cannot have it in that way, the event will be displatched for ALL installed captcha.

>  I would also deprecate the Captcha

This class to "rule them all", cannot be trashed.

Fedik committed:
Update

dgrammatiko commented:
@laoneo the registry concept is the correct approach not only for the Captcha but also for the Editor and the Auth plugins. All these plugins, as they are right now, dispatching events within an event which architecturally is wrong. What @Fedik is doing here is normalising to single event (if I understand the code correctly, haven't step through though) which is the right thing

laoneo commented:
So it means you can never have multiple captchas on one site?

dgrammatiko commented:
> So it means you can never have multiple captchas on one site?

No, you should be able to, as you can have multiple Editors, etc. The issue that Fedik is fixing here is that all these plugins basically used to be direct callbacks up to 3.x. J4 everything was converted to events but in reality the events were wrappers to the same methods (eg Editor->init()). The registry acts as event delegation, which imo is a good approach and a solid concept that will lead to less buggy code

Fedik commented:
hmhm, please look the code. That to much to explain :wink: 

You can have many captcha, and many differen captchas on same page.

But you cannot use `onDisplay` event. Please follow linked issue.

Fedik commented:
> I would also deprecate the Captcha

On second thought, maybe we could, and use Registry directly. 
But we do not have a clear way to access a "shared" class instance.
```php
$registry = Factory::getContainer()->get(CaptchaRegistry::class);
$captcha = $registry->get('potato_captcha');

echo $captcha->display();
```
So maybe in joomla 8 :wink:
Look on the Captcha class as a comfortable helper.

Fedik committed:
Update

Fedik committed:
Sync. Merge branch '4.4-dev' into captcha-event2

 Conflicts:
	plugins/captcha/recaptcha_invisible/src/Extension/InvisibleReCaptcha.php
	tests/Unit/Plugin/Captcha/InvisibleRecaptcha/Extension/InvisibleRecaptchaPluginTest.php

laoneo commented:
I guess it is better to rebase this pr to the 5.0-dev branch, as it will get more attention for new features. Thanks for understanding.

Fedik committed:
Merge branch '5.0-dev' into captcha-event2

Fedik committed:
Event improve

Fedik committed:
Event improve

Fedik committed:
Fixes

Fedik committed:
Better naming

Fedik committed:
Sync. Merge branch '5.0-dev' into captcha-event2

 Conflicts:
	tests/Unit/Plugin/Captcha/InvisibleRecaptcha/Extension/InvisibleRecaptchaPluginTest.php

Fedik committed:
phpcs

HLeithner committed:
Merge branch '5.0-dev' into captcha-event2

Fedik committed:
Merge branch '5.0-dev' into captcha-event2

HLeithner commented:
thanks
